--- a/extra/Configs/Config.in.arch
+++ b/extra/Configs/Config.in.arch
@@ -118,6 +118,21 @@ config ARCH_USE_MMU
 
 	  If you're unsure, answer Y.
 
+choice
+	prompt "uClibc Page Size"
+	default UCLIBC_PAGE_SIZE_4KB
+	help
+	  Select the uClibc page size
+
+	config UCLIBC_PAGE_SIZE_4KB
+		bool "4kB"
+	config UCLIBC_PAGE_SIZE_16KB
+		bool "16kB"
+	config UCLIBC_PAGE_SIZE_64KB
+		bool "64kB"
+
+endchoice
+
 config UCLIBC_HAS_FLOATS
 	bool "Enable floating point number support"
 	default y
--- a/libc/sysdeps/linux/mips/bits/uClibc_page.h
+++ b/libc/sysdeps/linux/mips/bits/uClibc_page.h
@@ -24,11 +24,17 @@
  * happens to have been configured.  It might use 4KB, 16K or 64K
  * pages.  To avoid using the current kernel configuration settings,
  * uClibc will simply use 4KB on mips and call it good. */
-#if 0
+
+/* define UCLIBC_PAGE_SIZE_16KB or UCLIBC_PAGE_SIZE_64KB from config */
+
+#ifdef __UCLIBC_PAGE_SIZE_64KB__
 #define PAGE_SHIFT	16
+#elif defined(__UCLIBC_PAGE_SIZE_16KB__)
 #define PAGE_SHIFT	14
-#endif
+#else
 #define PAGE_SHIFT	12
+#endif
+
 #define PAGE_SIZE	(1UL << PAGE_SHIFT)
 #define PAGE_MASK	(~(PAGE_SIZE-1))
 
