diff -Nur samba-3.0.24.org/source/auth/auth_sam.c samba-3.0.24/source/auth/auth_sam.c
--- samba-3.0.24.org/source/auth/auth_sam.c	2007-02-05 02:59:14.000000000 +0800
+++ samba-3.0.24/source/auth/auth_sam.c	2013-09-17 14:14:53.937759167 +0800
@@ -30,7 +30,15 @@
  Do a specific test for an smb password being correct, given a smb_password and
  the lanman and NT responses.
 ****************************************************************************/
-
+upercase_name(char * name)
+{
+	int i=0;
+	while (name[i]!='\0')
+	{
+		name[i]=toupper(name[i]);
+		i++;
+	}
+}
 static NTSTATUS sam_password_ok(const struct auth_context *auth_context,
 				TALLOC_CTX *mem_ctx,
 				struct samu *sampass, 
@@ -41,6 +49,7 @@
 	uint32 acct_ctrl;
 	const uint8 *lm_pw, *nt_pw;
 	const char *username = pdb_get_username(sampass);
+	NTSTATUS ret;
 
 	acct_ctrl = pdb_get_acct_ctrl(sampass);
 	if (acct_ctrl & ACB_PWNOTREQ) {
@@ -55,14 +64,31 @@
 
 	lm_pw = pdb_get_lanman_passwd(sampass);
 	nt_pw = pdb_get_nt_passwd(sampass);
+	
+ 
 
-	return ntlm_password_check(mem_ctx, &auth_context->challenge, 
+	ret= ntlm_password_check(mem_ctx, &auth_context->challenge, 
+				   &user_info->lm_resp, &user_info->nt_resp, 
+				   &user_info->lm_interactive_pwd, &user_info->nt_interactive_pwd,
+				   username, 
+				   user_info->smb_name,
+				   user_info->client_domain,
+				   lm_pw, nt_pw, user_sess_key, lm_sess_key);
+	if(!NT_STATUS_IS_OK(ret)){
+		//Fixed can't loging with Win7. The user name should be uppercase.
+		upercase_name(user_info->smb_name);
+		return ntlm_password_check(mem_ctx, &auth_context->challenge, 
 				   &user_info->lm_resp, &user_info->nt_resp, 
 				   &user_info->lm_interactive_pwd, &user_info->nt_interactive_pwd,
 				   username, 
 				   user_info->smb_name,
 				   user_info->client_domain,
 				   lm_pw, nt_pw, user_sess_key, lm_sess_key);
+		
+	}
+	else
+		return ret;
+	
 }
 
