--- radvd-1.8.3.old/send.c	2011-10-06 23:43:41.000000000 +0800
+++ radvd-1.8.3/send.c	2015-03-06 15:33:57.533949500 +0800
@@ -244,7 +244,9 @@
 
 			if (iface->cease_adv && prefix->DeprecatePrefixFlag) {
 				/* RFC4862, 5.5.3, step e) */
-				pinfo->nd_opt_pi_valid_time	= htonl(MIN_AdvValidLifetime);
+				//pinfo->nd_opt_pi_valid_time	= htonl(MIN_AdvValidLifetime);
+				// set valid_time to 0 requested by telenor
+				pinfo->nd_opt_pi_valid_time	= 0;
 				pinfo->nd_opt_pi_preferred_time = 0;
 			} else {
 				if (prefix->DecrementLifetimesFlag) {
